src/lib/job.rs

```diff
@@ -71,17 +86,44 @@ pub async fn process_job(job_id: &Uuid, jobs: Arc<RwLock<JobRegistry>>) {
         job.status = JobStatus::Running;
     }
 
+    let copy_job : JobDesc;
     {
         let job = &*job.read().await;
-        if let Err(e) = start_build_job(job).await {
-            error!("job {} failed due to: {}", &job_id, e);
-        } else {
-            succeed = true;
-        }
+
+        /*
+           Important: We can't pass job directly into start_build_job, otherwise we will
+           have a deadlock. We want to update the JobDesc for the job with the PID of the
+           running process, but since we have the read lock, we can't acquire the write lock.
+
+           If we want to avoid having to request the read lock everytime we need it, we should
```

Comment:
What about instead of a deep copy we just clone the `Arc<RwLock<JobDesc>>`, i know it's a bit boilerplate but we went through all these trouble just so we can avoid deep copy. I think current code works good so it's just my personal preference. Also implementing a macro can probably reduce the repetitive locking/unlocking.

@arora-aman  thoughts ?

Reply:
If we are doing a deep copy then we should probably do it earlier and write some other way of returning the status. But the way code is currently written it makes more sense to just clone the `Arc` instead of the cloning the underlying `JobDesc`

Reply:
I am not sure if that really is needed.
Things to keep in mind, why make it more complicated right now when the job is only used for information so far, which should technically not be modified (except for PID, but let's ignore that for now).
Second, the JobRegistry is now being passed around as well as the running job uuid, so technically no need to pass the `Arc<RwLock<JobDesc>>` cause we can get it from the `JobRegistry` directly.

Reply:
if u clone the Arc it allows for you to modify the Desc without need to hold lock over the registry

Reply:
I'm not 100% faimilar with this codebase so I could be wrong, but start_build_job(job.clone(), ..) seems simpler, and would work.

Reply:
hmm no, cloning the Arc still require the read lock

Reply:
So I am not doing anything about this. Like the comment I left in the code is saying, I don't think it is worth adding more concurrency complexity with the lock when the data of JobDesc is not expected to change. A comment reflects this decision, if we want to change that we should be making a new issue.